relayer/crates/starknet-integration-tests/src/tests/ics20.rs

```diff
@@ -589,16 +595,308 @@ fn test_starknet_ics20_contract() -> Result<(), Error> {
+
+            let response = starknet_chain.send_message(message).await?;
+
+            info!("ICS20 send packet response: {:?}", response);
+
+            let mut ibc_packet_events: Vec<PacketRelayEvents> =
+                event_encoding.filter_decode_events(&response.events)?;
+
+            info!("IBC packet events: {:?}", ibc_packet_events);
+
+            let mut ibc_transfer_events: Vec<IbcTransferEvent> =
+                event_encoding.filter_decode_events(&response.events)?;
+
+            info!("IBC transfer events: {:?}", ibc_transfer_events);
+
+            assert_eq!(ibc_packet_events.len(), 1);
+            assert_eq!(ibc_transfer_events.len(), 1);
+
+            let Some(PacketRelayEvents::Send(send_packet_event)) = ibc_packet_events.pop() else {
+                panic!("expected send packet event");
+            };
+
+            let Some(IbcTransferEvent::Send(send_ics20_event)) = ibc_transfer_events.pop() else {
+                panic!("expected send ics20 event");
+            };
+
+            (send_packet_event, send_ics20_event)
+        };
+
+        // create ibc packet
+
+        let starknet_ibc_packet = {
+            let timeout_height_on_b = IbcHeight::new(
+                send_packet_event.timeout_height_on_b.revision_number,
+                send_packet_event.timeout_height_on_b.revision_height,
+            )
+            .map(TimeoutHeight::At)
+            .unwrap_or_else(|_| TimeoutHeight::Never);
+
+            let timeout_timestamp_on_b = (send_packet_event.timeout_timestamp_on_b.timestamp > 0)
+                .then(|| {
+                    TimeoutTimestamp::At(IbcTimestamp::from_nanoseconds(
+                        send_packet_event.timeout_timestamp_on_b.timestamp * 1_000_000_000,
+                    ))
+                })
+                .unwrap_or(TimeoutTimestamp::Never);
+
+            let ibc_transfer_packet_data = PacketData {
+                token: PrefixedCoin {
+                    denom: send_ics20_event.denom.to_string().parse()?,
+                    amount: send_ics20_event.amount.to_string().parse()?,
+                },
+                sender: send_ics20_event.sender.to_string().into(),
+                receiver: send_ics20_event.receiver.to_string().into(),
+                memo: send_ics20_event.memo.into(),
+            };
+
+            IbcPacket {
+                seq_on_a: send_packet_event.sequence_on_a.sequence.into(),
+                port_id_on_a: send_packet_event.port_id_on_a.port_id.parse()?,
+                chan_id_on_a: send_packet_event.channel_id_on_a.channel_id.parse()?,
+                port_id_on_b: send_packet_event.port_id_on_b.port_id.parse()?,
+                chan_id_on_b: send_packet_event.channel_id_on_b.channel_id.parse()?,
+                data: serde_json::to_vec(&ibc_transfer_packet_data).unwrap(),
+                timeout_height_on_b,
+                timeout_timestamp_on_b,
+            }
+        };
```

Comment:
I wonder if `hermes-sdk` does not have a component that listens to send transfer events, handles these underlying insertions, and packet creation? Should we open an issue to add this to the hermes-sdk (or maybe the relayer, not entirely sure)?

Reply:
Yes, hermes-sdk can support this. I am using the `CanIbcTransferToken<CounterpartyChain>::ibc_transfer_token` when transfering token from cosmos-sdk to starknet.

https://github.com/informalsystems/ibc-starknet/blob/c13b888f907494910de113845fbbe559d50112bf/relayer/crates/starknet-integration-tests/src/tests/ics20.rs#L381-L393

But I don't know how to implement it for starknet to cosmos. We need to the following trait bound to be satisfied,

https://github.com/informalsystems/ibc-starknet/blob/c13b888f907494910de113845fbbe559d50112bf/relayer/crates/starknet-chain-context/src/contexts/chain.rs#L367-L368

Reply:
opened #181

Reply:
I get a bit worried about TODOs in the code. They’re like socks in the laundry, just disappearing forever!
Let’s create issues instead of leaving them there. To keep things simple, the issues don’t have to be super detailed. Just a short, straightforward one is fine unless extra details are really necessary.

Reply:
Got it, thanks for pointing that out!

Reply:
Oh, I see you already opened up a bunch. Much appreciated!